stdlib/builtins.pyi

```diff
@@ -97,23 +97,25 @@ class object:
     def __class__(self, __type: Type[object]) -> None: ...  # type: ignore # noqa: F811
     def __init__(self) -> None: ...
     def __new__(cls: Type[_T]) -> _T: ...
-    def __setattr__(self, name: str, value: Any) -> None: ...
-    def __eq__(self, o: object) -> bool: ...
-    def __ne__(self, o: object) -> bool: ...
+    def __setattr__(self, __name: str, __value: Any) -> None: ...
+    def __eq__(self, __o: object) -> bool: ...
+    def __ne__(self, __o: object) -> bool: ...
     def __str__(self) -> str: ...
     def __repr__(self) -> str: ...
     def __hash__(self) -> int: ...
-    def __format__(self, format_spec: str) -> str: ...
-    def __getattribute__(self, name: str) -> Any: ...
-    def __delattr__(self, name: str) -> None: ...
+    def __format__(self, __format_spec: str) -> str: ...
+    def __getattribute__(self, __name: str) -> Any: ...
+    def __delattr__(self, __name: str) -> None: ...
     def __sizeof__(self) -> int: ...
     def __reduce__(self) -> str | Tuple[Any, ...]: ...
     if sys.version_info >= (3, 8):
-        def __reduce_ex__(self, protocol: SupportsIndex) -> str | Tuple[Any, ...]: ...
+        def __reduce_ex__(self, __protocol: SupportsIndex) -> str | Tuple[Any, ...]: ...
     else:
-        def __reduce_ex__(self, protocol: int) -> str | Tuple[Any, ...]: ...
+        def __reduce_ex__(self, __protocol: int) -> str | Tuple[Any, ...]: ...
     def __dir__(self) -> Iterable[str]: ...
     def __init_subclass__(cls) -> None: ...
+    @classmethod
+    def __subclasshook__(cls, *args: object) -> NotImplementedType: ...
```

Comment:
I don't think the type system supports `NotImplementedType`. This should return bool or NotImplemented, right? I think in that case we should just say `bool` and type checkers need to know to special case `NotImplemented` in dunders.

Reply:
``object.__subclasshook__`` can be guaranteed to only return ``NotImplemented``, no matter what you pass it, and no matter how many arguments you pass it, as long as they are positional arguments. I tried annotating it as return ``Literal[NotImplemented], but mypy didn't like it. It was much happier with ``NotImplementedType`` -- the reason for the CI failure is because in "real Python", ``NotImplementedType`` is defined in the ``types`` module, whereas typeshed gives the type as living in ``builtins``.

Reply:
But yes, for custom classes that override ``__subclasshook__``, the method should only return ``bool`` or ``NotImplemented``.

Reply:
Yes, `Literal[NotImplemented]` is illegal, you can only have literal strs, enums, ints, or bools according to PEP 586.

Reply:
Tbh I think I'll just cut out ``__subclasshook__`` for now and leave it for another PR. It's an interesting question, but one for another day.